Support for multiple initial populations

mgramigna opened the pull request:
# Summary

Fixes #149 

This PR enables the results processing logic of `fqm-execution` to handle the case of multiple Initial Populations defined in a measure. Note that this is only applicable for Ratio Measures.

For context, in the case of a ratio measure when there are two IPs, the numerator and the denominator are treated independently when they depend on different IPs. 

## New behavior

* Ratio measures with 2 IPs no longer have numerator/denominator dependencies
  * Results for each IP are reported in populationResults and episodeResults where appropriate
* Population relevance has special case handing for multiple IP measures
* Results and relevance in these cases are determined by looking at the `criteriaReference` extension in the numer/denom population
* Episode-based measures can now be detected by the `cqfm-populationBasis` extension being present at the individual population level (as opposed to Measure or Measure.group)

## Code changes

* Update clause results builder for relevants/results post-processing to only have the numerator and denominator influence each other when we are _not_ in a multiple-ipp case
* Update all the results processing to use `criteriaExpression` when provided to resolve ambiguity between multiple populations of the same type
* Update episode of care group checking to have special handling for ratio measures, allowing populationBasis to be specified at the population level
* Unify all uses of checking for scoringCode to helper functions
* Add tests for everything
* Unify the two different MeasureBundleHelper test files into 1 

# Testing guidance

1. Grab the bundle and patient from #149 and run the CLI, ensuring that the results are as expected:

```
npm run cli -- detailed -m measure-bundle.json -p patient-bundle.json
```

Verify the differences with the new output and what was there on master before. You should see that `ipp2` is now properly being reported as `true`

2. Download [multiple-ipp-tests.zip](https://github.com/projecttacoma/fqm-execution/files/9912461/multiple-ipp-tests.zip) and extract the individual files within it to the root of `fqm-execution`. These are test cases I created and bundled that further test this logic, specifically testing the "independentness" of numer and denom in these multiple IPP cases. Instructions are in `MULTIPLE-IP-INSTRUCTIONS.md`, but it's a two parter:

**Part 2A: Test cases**

Run the `./run-multiple-ip-test-cases.sh` file to generate 3 JSON outputs for each of the CQL libraries present in the `multiple-ipp-test-cases` directory. Please cross-reference the results with the CQL files.

* `false-denom-output.json` - Of note in this is that `ipp` and `denom` both have no results, but `ipp2` and `numer` have one episode result. you should see that the results for `ipp2` and `numer` are reported as true independent of `ipp` and `denom`. This would not happen on master. In addition, `denom` is not relevant, as its corresponding ipp is falsy. `numerator` is relevant though, since its corresponding ipp is truthy
* `false-numer-output.json` - Of note in this one is that the numerator population is not relevant since its ipp is false, but denom/ipp are both relevant and truthy
* `both-true-output.json` - Everything is true and relevant, and we get results for 2 episodes

**Part 2B: Regression Tests**

Given that a lot of the results processing has changed in this PR, I wrote some scripts that ensure that previous calculation behavior for measures without two IPPs is unchanged.

* Clone connectathon to the root of fqm-execution if you don't have it: `git clone https://github.com/DBCG/connectathon.git`
* Run the regression script: `./run-regression.sh`

`PASS` means that there are 0 diffs in the detailed results for that measure/test case between master and this branch. This is expected, as none of the measures in connectathon have two IPPs, so calculation results should be unchanged.

elsaperelli reviewed (changes requested):
Nice! Just some small changes and questions. My only overall suggestion is to maybe add some unit tests for DetailedResultsBuilder.ts since there were some changes there that are not covered? Unsure how necessary that is though with all the external tests you made.

mgramigna commented on src/calculation/ClauseResultsBuilder.ts:
```diff
@@ -642,7 +645,25 @@ export function buildPopulationRelevanceMap(results: PopulationResult[]): Popula
+    if (numerRelevantIPP) {
+      if (getResult(PopulationType.IPP, results, numerRelevantIPP.criteria.expression) === false) {
+        if (hasResult(PopulationType.NUMER, relevantResults)) {
+          setResult(PopulationType.NUMER, false, relevantResults);
```
I believe so... good call!

mgramigna commented on src/calculation/ClauseResultsBuilder.ts:
```diff
@@ -642,7 +645,25 @@ export function buildPopulationRelevanceMap(results: PopulationResult[]): Popula
+    if (denomRelevantIPP) {
+      if (getResult(PopulationType.IPP, results, denomRelevantIPP.criteria.expression) === false) {
+        if (hasResult(PopulationType.DENOM, relevantResults)) {
+          setResult(PopulationType.DENOM, false, relevantResults);
```
I believe so... good call!

mgramigna commented:
> Nice! Just some small changes and questions. My only overall suggestion is to maybe add some unit tests for DetailedResultsBuilder.ts since there were some changes there that are not covered? Unsure how necessary that is though with all the external tests you made.

I Added some more tests for the branch of logic that I changed in `handlePopuationValues` of `DetailedResultsBuilder`. I didn't go full crazy and add tests for everything in this function since it was completely untested (I was tempted, but I think it's a little out of scope for this)

Ideally we can expand the tests more in the future

elsaperelli reviewed (approved):
Lgtm! Thank you for adding the detailed comments! Those will be helpful. As well as the additional tests.

hossenlopp reviewed (commented):
From a quick pass, the approach is sound. 👍 

I didn't see anything relevant to making sure the distinct IPPs are getting results in the MeasureReportBuilder. Maybe it already is inherently doing it right.

mgramigna commented:
> From a quick pass, the approach is sound. 👍
> 
> I didn't see anything relevant to making sure the distinct IPPs are getting results in the MeasureReportBuilder. Maybe it already is inherently doing it right.

Yeah, I think we will have the same problem as we do with ratio measures with multiple observations: need to differentiate them in the MeasureReport. We're avoiding this for now. If [this ticket](https://jira.hl7.org/browse/FHIR-38832) gets in, we can probably use this to differentiate the populations. For now, I wonder if (as a separate PR) we should update the MeasureReportBuilder to either throw an error or a warning or something when we encounter one of these edge cases, just so users don't expect it to work (and we can recommend using the low-level detailed results API or something)

natjoe4 commented on test/ClauseResultsBuilder.test.ts:
```diff
@@ -428,5 +462,269 @@ describe('ResultsHelper', () => {
+            },
+            criteria: {
+              language: 'text/cql',
+              expression: 'ipp'
```
I think we need this to match the valueString in the criteriaReference extension in denominator for our getResult code to work

mgramigna replied:
The valueString in that extension refers to the `id` of the referenced population, not the criteria expression.

natjoe4 commented on test/ClauseResultsBuilder.test.ts:
```diff
@@ -428,5 +462,269 @@ describe('ResultsHelper', () => {
+            },
+            criteria: {
+              language: 'text/cql',
+              expression: 'ipp2'
```
Here too for numerator

mgramigna replied:
The valueString in that extension refers to the `id` of the referenced population, not the criteria expression.

natjoe4 reviewed (changes requested):
Tested thoroughly and works well! Took a good look at the code diff and only have a couple tiny comments. Nice work!
